demo/rust/micdemo/src/main.rs

```diff
@@ -0,0 +1,299 @@
+/*
+    Copyright 2018-2021 Picovoice Inc.
+
+    You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
+    file accompanying this source.
+
+    Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+    an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+    specific language governing permissions and limitations under the License.
+*/
+
+use clap::{App, Arg, ArgGroup};
+use ctrlc;
+use miniaudio;
+use pv_porcupine::{pv_keyword_paths, pv_library_path, pv_model_path, Porcupine, KEYWORDS};
+use std::path::PathBuf;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::time::SystemTime;
+
+static LISTENING: AtomicBool = AtomicBool::new(true);
+
+fn process_audio_chunk(
+    samples: &[i16],
+    porcupine: &mut Porcupine,
+    buffer: &mut Vec<i16>,
+    keywords: &[String],
+) {
+    buffer.extend_from_slice(samples);
+
+    while buffer.len() >= porcupine.frame_length() as usize && LISTENING.load(Ordering::SeqCst) {
```

Comment:
I think it's possible we could get a smaller buffer in a rare case. Best not to exit the recording loop, but just report that we got a small buffer, don't pass it to porcupine and continue to loop. If we get a larger buffer then frame_length, report it to console, pass as many subframes of it to porcupine as you can and then continue the loop.

Reply:
That is what the `buffer` is for, to handle different sized inputs. Unless I'm missing an edge case I don't see?

Reply:
ah okay, so buffer is always a multiple of porcupine frame length? if there is any possibility that may not be true we should handle it. If not, great!

Reply:
I think what you are thinking of the `buffer` is actually the `samples`. The `samples` is what is returned to us from the audio library callback. We add `samples` another buffer `buffer`, and then while `buffer` is longer than a single frame, we "drain" (popping iterator) the buffer `frame_length` items. `buffer` persists across callback calls, so whether the `samples` is larger or smaller than a single `frame_length` doesn't matter.